### ISSUE TITLE
No lexical declarations inside CaseClause StatementList 

### ISSUE THREAD

**Issue by arv** [eslint/eslint#4278]
The following code has bad semantics:

```js
let x = 2;
switch (2) {
  case 1:
    let x = 42;
    break;
  case 2:
    let y = x;  // TDZ error for x since it was never initialized.
   break;
}
```

because let/const (and class and functions) are scoped to the switch block.

It would be nice to have a lint rule saying that let, const, class and function are not allowed (as direct children) in the StatementList of a CaseClause (and same for DefaultClause).

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by arv** [comment]
It seems easy enough to write. I'll take a look.

**Comment by nzakas** [comment]
We already catch this issue via `no-redeclare`

**Comment by arv** [comment]
That is a different issue. That was a bad example. Let me try to clarify.

The problem is that lexical declarations are "hoisted" to the nearest block which is the SwitchStatement. The problem is that the CaseClause is not a block in JS. Here is another example:

```js
switch (2) {
  case 1:
    let x = 1;
    print(x);
    break;
  case 2:
    let y = x;  // This always fails.
    print(y);
    break;
 }
```

The right way to write this is:

```js
switch (2) {
  case 1: {
    let x = 1;
    print(x);
    break;
  }
  case 2: {
    let y = x;  // eslint: x is undefined
    print(y);
    break;
  }
 }
```

**Comment by arv** [comment]
I implemented this.

https://github.com/eslint/eslint/compare/master...arv:no-case-declarations

I guess I should just have created a plugin in the first place.

**Comment by mysticatea** [comment]
Thank you for reporting, @arv .

```js
switch (a) {
  case 0:
    let x = 0;
    break;

  case 1:
    foo(x); // This `x` always throws a runtime error because of TDZ.
    break;
}
```

I felt this should be reported with eslint.
I think there are some directions:

- It forbids declarations in CaseClause.
- It reports references of variables which are declared in other CaseClause. (enhancement of `no-undef` or `no-use-before-define`?)

`no-case-declarations` looks good to me.

**Comment by michaelficarra** [comment]
:+1: please reopen.

**Comment by arv** [comment]
@nzakas Do you want me to create a PR for this or do you want the discussion to settle down a bit first?

**Comment by nzakas** [comment]
Thanks for explaining. Go ahead and create a PR.
